pyccel/ast/numpyext.py

```diff
@@ -1201,8 +1224,38 @@ def __init__(self, x):
         self._precision = default_precision[str_dtype(self._dtype)]
 
 class NumpyMod(NumpyUfuncBinary):
-    def __init__(self, x1, x2):
-        self.copy_attributes(x1)
+    def __init__(self, *args):
+        integers  = [a for a in args if a.dtype is NativeInteger() or a.dtype is NativeBool()]
+        reals     = [a for a in args if a.dtype is NativeReal()]
+        others    = [a for a in args if a not in integers+reals]
+
+        if others:
+            raise TypeError('{} not supported'.format(others[0].dtype))
+
+        if reals:
+            self._dtype     = NativeReal()
+            self._precision = max(a.precision for a in reals)
+        elif integers:
+            self._dtype     = NativeInteger()
+            self._precision = max(a.precision for a in integers)
+        else:
+            raise TypeError('cannot determine the type of {}'.format(self))
+
+        shapes = [a.shape for a in args]
+
+        if all(sh is not None for sh in shapes):
+            if len(args) == 1:
+                shape = args[0].shape
+            else:
+                shape = broadcast(args[0].shape, args[1].shape)
+
+                for a in args[2:]:
+                    shape = broadcast(shape, a.shape)
+
+            self._shape = shape
+            self._rank  = len(shape)
+        else:
+            self._rank = max(a.rank for a in args)
```

Comment:
This code looks sufficiently general to me, that we should be able to move it to the `__init__` of class `NumpyUfuncBinary`.

Reply:
it depends on the function no ?, for example in NumpyMod we don't take complex arguments

Reply:
I think we can consider the complex case and let the Python interpreter raise an error if that is not allowed...

Reply:
you have arctan2 for example can take two integers and return a real number, contrary to mod where it returns an integer if he gets integers and a real if he gets reals

Reply:
We could make this a function where `complexes`, `reals`, etc are initialized unless an argument is provided in which case `complexes = []`. Then a second argument for the minimum output precision.
e.g. int+int would normally give int but with minimum precision real it gives real

It might make for a slightly complicated function but at least we won't have to write out all 30 lines for each special case

Reply:
@EmilyBourne This is a good suggestion. In `NumpyUfuncBinary` we could add a static method:
```
@staticmethod
def _initialize(obj, arg1, arg2, allow_complex=False, basic_dtype='int'):
    ...
```
which gets called by the various `__init__` functions with the proper arguments. But how many binary functions do we have?

Reply:
As @saidctb pointed out in a private conversation, most "magic" binary functions need broadcasting in Numpy, e.g. `__add__`, `__mul__`, `__div__`, etc. This will be better addressed in a separate pull request.